# CRISP-DM Methodology: Data Preparation Stage
This notebook is part of the **Data Preparation** stage of the CRISP-DM methodology.

### Stage Overview:
Data Preparation is the process of cleaning, transforming, and organizing raw data into a format suitable for analysis and modeling.

The following tasks will be performed in this notebook:
1. **Handling Missing Values**: Identify and resolve missing data issues.
2. **Handling Duplicates**: Remove duplicate records from the dataset.
3. **Feature Engineering**: Modify or create new features for better model performance.
4. **Encoding Categorical Variables**: Transform categorical data into numerical formats.
5. **Data Normalization/Scaling**: Standardize numerical features for consistency.


<h1>Data Preprocessing on leetcode's scraped data using Python, Numpy & Pandas</h1>

<center>
    <img src="https://i.imgur.com/FGODlma.jpg" width=60% style="box-shadow:rgba(52, 64, 77, 0.2);border-radius:10px;">
</center>

#### `Introduction`

- This notebook revolves around the comprehensive data preprocessing and feature engineering of a dataset containing 3000 LeetCode problems.
- In this dataset, 840 of these problems are classified as premium, which posed a unique challenge due to a lack of additional data.

<center><h1> Overview of the Scraped Data</h1></center>

In [1]:
!pip install pandas numpy

In [2]:
# Import necessary libraries
import numpy as np
import pandas as pd

# Suppress warnings to enhance the code output clarity
import warnings
warnings.filterwarnings("ignore")

In [4]:
# Read the scraped data from the CSV file into a DataFrame
df = pd.read_csv('leetcode_scraped_data.csv')
dff = df.copy()

In [5]:
# Display a random sample of 10 rows from the DataFrame
df.sample(10)

title  \
1709                     1710. Maximum Units on a Truck   
1890                              1891. Cutting Ribbons   
1969           1970. Last Day Where You Can Still Cross   
965                             966. Vowel Spellchecker   
805                806. Number of Lines To Write String   
663                                664. Strange Printer   
1233    1234. Replace the Substring for Balanced String   
2374     2375. Construct Smallest Number From DI String   
1793  1794. Count Pairs of Equal Substrings With Min...   
675                     676. Implement Magic Dictionary   

                                            problem_URL acceptance difficulty  \
1709  https://leetcode.com/problems/maximum-units-on...      73.9%       Easy   
1890      https://leetcode.com/problems/cutting-ribbons      51.1%     Medium   
1969  https://leetcode.com/problems/last-day-where-y...      61.9%       Hard   
965    https://leetcode.com/problems/vowel-spellchecker      51.4%     Medium   
805   https://leetcode.com/problems/number-of-lines-...      69.7%       Easy   
663       https://leetcode.com/problems/strange-printer      61.0%       Hard   
1233  https://leetcode.com/problems/replace-the-subs...      38.8%     Medium   
2374  https://leetcode.com/problems/construct-smalle...      76.2%     Medium   
1793  https://leetcode.com/problems/count-pairs-of-e...      65.2%     Medium   
675   https://leetcode.com/problems/implement-magic-...      57.8%     Medium   

                                           solution_URL is_premium  \
1709  https://leetcode.com/problems/maximum-units-on...      False   
1890  https://leetcode.com/problems/cutting-ribbons/...        NaN   
1969  https://leetcode.com/problems/last-day-where-y...      False   
965   https://leetcode.com/problems/vowel-spellcheck...      False   
805   https://leetcode.com/problems/number-of-lines-...      False   
663   https://leetcode.com/problems/strange-printer/...      False   
1233  https://leetcode.com/problems/replace-the-subs...      False   
2374  https://leetcode.com/problems/construct-smalle...      False   
1793  https://leetcode.com/problems/count-pairs-of-e...        NaN   
675   https://leetcode.com/problems/implement-magic-...      False   

                                    problem_description  \
1709  You are assigned to put some amount of boxes o...   
1890                                                NaN   
1969  There is a 1-based binary matrix where 0 repre...   
965   Given a wordlist, we want to implement a spell...   
805   You are given a string s of lowercase English ...   
663   There is a strange printer with the following ...   
1233  You are given a string s of length n containin...   
2374  You are given a 0-indexed string pattern of le...   
1793                                                NaN   
675   Design a data structure that is initialized wi...   

                                             topic_tags accepted submission  \
1709                       'Array', 'Greedy', 'Sorting'   326.7K       442K   
1890                                                NaN      NaN        NaN   
1969  'Array', 'Binary Search', 'Depth-First Search'...    58.9K      95.1K   
965                     'Array', 'Hash Table', 'String'    43.7K      85.1K   
805                                   'Array', 'String'    89.5K     128.5K   
663                     'String', 'Dynamic Programming'   149.1K     244.6K   
1233                         'String', 'Sliding Window'    37.6K      96.9K   
2374        'String', 'Backtracking', 'Stack', 'Greedy'      37K      48.6K   
1793                                                NaN      NaN        NaN   
675   'Hash Table', 'String', 'Depth-First Search', ...    89.3K     154.4K   

     discussion_count likes                                  similar_questions  
1709               18  3.9K         'Maximum Bags With Full Capacity of Rocks'  
1890              NaN   NaN                            

In [6]:
# Display concise information about the DataFrame, including the data types and memory usage
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   title                3000 non-null   object
 1   problem_URL          3000 non-null   object
 2   acceptance           3000 non-null   object
 3   difficulty           3000 non-null   object
 4   solution_URL         3000 non-null   object
 5   is_premium           2400 non-null   object
 6   problem_description  2400 non-null   object
 7   topic_tags           2348 non-null   object
 8   accepted             2400 non-null   object
 9   submission           2400 non-null   object
 10  discussion_count     2400 non-null   object
 11  likes                2400 non-null   object
 12  similar_questions    1605 non-null   object
dtypes: object(13)
memory usage: 304.8+ KB


In [7]:
# Count the number of duplicated rows in the DataFrame
df.duplicated().sum()

0

In [8]:
# Count the number of missing values in each column of the DataFrame
df.isnull().sum()

title                     0
problem_URL               0
acceptance                0
difficulty                0
solution_URL              0
is_premium              600
problem_description     600
topic_tags              652
accepted                600
submission              600
discussion_count        600
likes                   600
similar_questions      1395
dtype: int64

- We have data of 3000 LeetCode problems.
- Out of these, 600 are premium problems, so we lack additional data for them.
- There are 652 null values in the `topic_tags` column. This is because 600 premium problems and other 52 null values indicate that the topic tags information is not scrape.
- In the `similar_questions` column, there are 1,395 null values. This is because 600 premium problems and other 795 null values indicate that similar questions is not available in the leetcode website.

<center> <h2> Handling Missing Values </h2> </center>

> <p style="color:red"> Note: We lack information regarding premium questions. </p>

### `is_premium`

- Handle the missing values in the `is_premium` column by replacing them with the boolean value `True` to accurately denote the premium status.

In [9]:
# Get the unique values in the `is_premium` column.
df['is_premium'].unique()

array([False, nan], dtype=object)

In [10]:
#  Check the number of missing values in the `is_premium` column.
df['is_premium'].isnull().sum()

600

In [11]:
# Replace all missing values in the `is_premium` column with `True`.
df['is_premium'].fillna(True, inplace=True)

In [12]:
# Get the number of premium and non-premium questions in the DataFrame.
df['is_premium'].value_counts()

is_premium
False    2400
True      600
Name: count, dtype: int64

In [13]:
# Get the data type of the 'is_premium' column.
df['is_premium'].dtype

dtype('bool')

### `topic_tags`

- Fill the null values with the appropriate values, namely "JavaScript" and "pandas".

In [14]:
#  Check the number of missing values in the `topic_tags` column.
df['topic_tags'].isnull().sum()

652

In [15]:
# The `selected_ids` list contains a set of LeetCode question IDs
# that are all related to JavaScript
# Number of `selected_ids` is 37
selected_ids = [2693, 2694, 2695, 2703, 2704, 2705, 2715, 2721, 2722, 2723, 2724,
2725, 2726, 2727, 2618, 2619, 2620, 2621, 2622, 2623, 2624, 2625, 2626,
2627, 2629, 2630, 2631, 2634, 2635, 2637, 2648, 2649, 2650, 2665, 2666, 2667, 2677]

In [16]:
# Assign "JavaScript" topic tag to rows where title starts with selected IDs.
df.loc[df['title'].str.split('.').apply(lambda x: x[0]).astype('int').isin(selected_ids), 'topic_tags'] = "JavaScript"

In [17]:
# The `selected_ids` list contains a set of LeetCode question IDs
# that are all related to Pandas
# Number of `selected_ids` is 15
selected_ids = [2877, 2878, 2879, 2880, 2881, 2882, 2883, 2884, 2885, 2886, 2887,
2888, 2889, 2890, 2891]

In [18]:
# Assign "Pandas" topic tag to rows where title starts with selected IDs.
df.loc[df['title'].str.split('.').apply(lambda x: x[0]).astype('int').isin(selected_ids), 'topic_tags'] = "Pandas"

### `similar_questions`

- Fill in the null values for non-premium questions with an empty string, as there are no similar questions for those LeetCode problems.

In [20]:
# Count the number of null values in the 'similar_questions'
df['similar_questions'].isnull().sum()

1395

In [21]:
# Set an empty string for 'similar_questions' where 'is_premium' is False and 'similar_questions' is null
df.loc[(df['is_premium'] == False) & df['similar_questions'].isnull(), 'similar_questions'] = ''

In [22]:
# Count the number of null values in the 'similar_questions'
df['similar_questions'].isnull().sum()

600

> **We handle all missing values for non-premium questions.**

<center> <h2> Feature Engineering </h2> </center>

### `title`

- We are extracting the problem numbers (IDs) from the `title`, which can be useful for organizing and analyzing the data.

In [24]:
# Retrieve the 'title' column
df['title']

0                                              1. Two Sum
1                                      2. Add Two Numbers
2       3. Longest Substring Without Repeating Characters
3                          4. Median of Two Sorted Arrays
4                        5. Longest Palindromic Substring
                              ...                        
2995    2996. Smallest Missing Integer Greater Than Se...
2996    2997. Minimum Number of Operations to Make Arr...
2997    2998. Minimum Number of Operations to Make X a...
2998          2999. Count the Number of Powerful Integers
2999     3000. Maximum Area of Longest Diagonal Rectangle
Name: title, Length: 3000, dtype: object

In [25]:
# Extract the numerical part from the 'title' column by
# splitting at '.' and converting to integers
df['title'].str.split('.').apply(lambda x: x[0]).astype('int')

0          1
1          2
2          3
3          4
4          5
        ... 
2995    2996
2996    2997
2997    2998
2998    2999
2999    3000
Name: title, Length: 3000, dtype: int64

In [26]:
# Extract numeric values from 'title' column, convert to integers,
# and assign to a new 'id' column
df['id'] = df['title'].str.split('.').apply(lambda x: x[0]).astype('int')

### `page_number`

- page_number: Number of web pages ranges from 1 to 60. per page contains 50 problems, total 3000 problems.

In [27]:
df['page_number'] = pd.cut(df['id'], bins=range(1, 3002, 50), include_lowest=True, right=False).apply(lambda x: (x.left//50)+1)

### `similar_questions`

- **string splitting** on `similar_question`
- We are extracting a numerical feature from the `similar_questions` column that represents the number of similar questions `no_similar_questions`.

In [28]:
# Retrieve the 'similar_questions' column
df['similar_questions']

0       '3Sum', '4Sum', 'Two Sum II - Input Array Is S...
1       'Multiply Strings', 'Add Binary', 'Sum of Two ...
2       'Longest Substring with At Most Two Distinct C...
3                    'Median of a Row Wise Sorted Matrix'
4       'Shortest Palindrome', 'Palindrome Permutation...
                              ...                        
2995    'Longest Common Prefix', 'First Missing Positi...
2996                'Minimum Bit Flips to Convert Number'
2997    'Shortest Bridge', 'Minimum Moves to Spread St...
2998    'Powerful Integers', 'Numbers With Repeated Di...
2999                                                     
Name: similar_questions, Length: 3000, dtype: object

In [29]:
# Performing string manipulation on the 'similar_questions' column in the DataFrame 'df':
# Removing single quotes and then splitting the string into a list using ', ' as the delimiter.
df['similar_questions'].str.split(", ")

0       ['3Sum', '4Sum', 'Two Sum II - Input Array Is ...
1       ['Multiply Strings', 'Add Binary', 'Sum of Two...
2       ['Longest Substring with At Most Two Distinct ...
3                  ['Median of a Row Wise Sorted Matrix']
4       ['Shortest Palindrome', 'Palindrome Permutatio...
                              ...                        
2995    ['Longest Common Prefix', 'First Missing Posit...
2996              ['Minimum Bit Flips to Convert Number']
2997    ['Shortest Bridge', 'Minimum Moves to Spread S...
2998    ['Powerful Integers', 'Numbers With Repeated D...
2999                                                   []
Name: similar_questions, Length: 3000, dtype: object

In [30]:
# Removing single quotes from the 'similar_questions' column in the DataFrame 'df'
# and then splitting the string into a list using ', ' as the delimiter.
df['similar_questions'] = df['similar_questions'].str.split(", ")

In [31]:
# Applying a lambda function to the 'similar_questions' column in the DataFrame 'df':
# Counting the number of elements in each list (if it is a list and the first element is not an empty string), otherwise returning 0.
df['similar_questions'].apply(lambda x: len(x) if isinstance(x,list) and x[0]!='' else 0)

0       21
1        8
2       11
3        1
4        6
        ..
2995     3
2996     1
2997     2
2998     2
2999     0
Name: similar_questions, Length: 3000, dtype: int64

In [32]:
# Creating a new column 'no_similar_questions' in the DataFrame 'df':
# Assigning the count of elements in each list from the 'similar_questions' column, or 0 if it's not a list.
df['no_similar_questions'] = df['similar_questions'].apply(lambda x: len(x) if isinstance(x,list) and x!=[''] else 0)

In [33]:
# Setting 'no_similar_questions' to pd.NA for rows where 'is_premium' is True
# and 'no_similar_questions' is 0.
df.loc[(df['is_premium'] == True) & (df['no_similar_questions'] == 0), 'no_similar_questions'] = pd.NA

> **We have introduced several new features, including `no_similar_questions` for number of similar questions, an `id` attribute for unique identification & `page_number` denotes the specific page on the website where the LeetCode problem appears.**

<center> <h2> Data Format Conversion and Data Type Conversion </h2> </center>

- We are removing special characters like **%** and abbreviations like **K** and **M** to represent the values as numbers.
- We are changing the data type of the columns from object data type to float or int to make them compatible with numerical analysis.

In [35]:
# Retrieve specific columns ['acceptance', 'accepted', 'submission', 'solution', 'discussion_count', 'likes', 'dislikes']
# from the DataFrame 'df'
df[['acceptance','accepted','submission','discussion_count','likes']]

acceptance accepted submission discussion_count  likes
0         54.4%    15.4M      28.4M             1.1K  59.2K
1         44.8%     5.2M      11.7M              739  32.3K
2         35.9%     6.7M      18.6M              457  40.8K
3         42.2%       3M       7.2M              562  29.1K
4         34.9%     3.5M        10M              371    30K
...         ...      ...        ...              ...    ...
2995      33.1%    35.1K     105.9K               60    122
2996      85.9%   113.8K     132.5K              103    590
2997      47.2%    22.3K      47.3K               25    259
2998      26.4%     5.4K      20.3K                7    109
2999      36.4%    38.7K     106.3K               15    109

[3000 rows x 5 columns]

In [36]:
# Displaying information about selected columns ['acceptance', 'accepted', 'submission', 'solution', 'discussion_count', 'likes', 'dislikes']
# in the DataFrame 'df'.
df[['acceptance','accepted','submission','discussion_count','likes']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   acceptance        3000 non-null   object
 1   accepted          2400 non-null   object
 2   submission        2400 non-null   object
 3   discussion_count  2400 non-null   object
 4   likes             2400 non-null   object
dtypes: object(5)
memory usage: 117.3+ KB


### `acceptance`

In [37]:
# Stripping the percentage sign from the 'acceptance' column in the DataFrame 'df'
# and converting the values to floating-point numbers.
df['acceptance'] = df['acceptance'].str.rstrip('%').astype('float')

### `accepted`

In [38]:
# Replacing 'K' with 'e3' and 'M' with 'e6' in the 'accepted' column of the DataFrame 'df', then converting the values to floating-point numbers.
df['accepted'] = df['accepted'].str.replace('K', 'e3').str.replace('M', 'e6').astype(float)

# Converting the 'accepted' column to numeric values, handling errors by coercing non-numeric values to NaN, and then converting to nullable integer type (Int32).
df['accepted'] = pd.to_numeric(df['accepted'], errors='coerce').astype('Int32')

### `submission`

In [39]:
# Replacing 'K' with 'e3' and 'M' with 'e6' in the 'submission' column of the DataFrame 'df', then converting the values to floating-point numbers.
df['submission'] = df['submission'].str.replace('K', 'e3').str.replace('M', 'e6').astype(float)

# Converting the 'submission' column to numeric values, handling errors by coercing non-numeric values to NaN, and then converting to nullable integer type (Int32).
df['submission'] = pd.to_numeric(df['submission'], errors='coerce').astype('Int32')

### `discussion_count`

In [41]:
# Converting the 'discussion_count' column in the DataFrame 'df' to numeric values, handling errors by coercing non-numeric values to NaN,
# and then converting to nullable integer type (Int32).

df['discussion_count'] = pd.to_numeric(df['discussion_count'], errors='coerce').astype('Int32')

### `likes`

In [42]:
# Replacing 'K' with 'e3' and 'M' with 'e6' in the 'likes' column of the DataFrame 'df', then converting the values to floating-point numbers.
df['likes'] = df['likes'].str.replace('K', 'e3').str.replace('M', 'e6').astype(float)

# Converting the 'likes' column to numeric values, handling errors by coercing non-numeric values to NaN, and then converting to nullable integer type (Int32).
df['likes'] = pd.to_numeric(df['likes'], errors='coerce').astype('Int32')

In [43]:
# Displaying selected columns ['acceptance', 'accepted', 'submission', 'solution', 'discussion_count', 'likes', 'dislikes'] from the DataFrame 'df'

df[['acceptance','accepted','submission','discussion_count','likes']]

acceptance  accepted  submission  discussion_count  likes
0           54.4  15400000    28400000              <NA>  59200
1           44.8   5200000    11700000               739  32300
2           35.9   6700000    18600000               457  40800
3           42.2   3000000     7200000               562  29100
4           34.9   3500000    10000000               371  30000
...          ...       ...         ...               ...    ...
2995        33.1     35100      105900                60    122
2996        85.9    113800      132500               103    590
2997        47.2     22300       47300                25    259
2998        26.4      5400       20300                 7    109
2999        36.4     38700      106300                15    109

[3000 rows x 5 columns]

<center> <h2> Column Rearrangement </h2> </center>

- This process involves changing the sequence of columns to make the DataFrame more organized, intuitive and suitable for analysis.

In [44]:
# Displaying the shape (number of rows and columns) of the DataFrame 'df'
df.shape

(3000, 16)

In [45]:
# Arrangement of columns
column_order = [
    'id',
    'page_number',
    'is_premium',
    'title',
    'problem_description',
    'topic_tags',
    'difficulty',
    'similar_questions',
    'no_similar_questions',
    'acceptance',
    'accepted',
    'submission',
    'discussion_count',
    'likes',
    'problem_URL',
    'solution_URL'
]

In [46]:
# Check number of columns
len(column_order)

16

In [47]:
# Column Rearrangement
df = df[column_order]

### `Final Overview of the preprocessed data`

In [48]:
# Information about dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   id                    3000 non-null   int64   
 1   page_number           3000 non-null   category
 2   is_premium            3000 non-null   bool    
 3   title                 3000 non-null   object  
 4   problem_description   2400 non-null   object  
 5   topic_tags            2400 non-null   object  
 6   difficulty            3000 non-null   object  
 7   similar_questions     2400 non-null   object  
 8   no_similar_questions  2400 non-null   float64 
 9   acceptance            3000 non-null   float64 
 10  accepted              2400 non-null   Int32   
 11  submission            2400 non-null   Int32   
 12  discussion_count      2399 non-null   Int32   
 13  likes                 2400 non-null   Int32   
 14  problem_URL           3000 non-null   object  
 15  solu

In [49]:
# Displaying the count of null values for columns with non-zero null counts
# in the DataFrame 'df'

df.isnull().sum()[df.isnull().sum()!=0]

problem_description     600
topic_tags              600
similar_questions       600
no_similar_questions    600
accepted                600
submission              600
discussion_count        601
likes                   600
dtype: int64

- We lack some information regarding 840 premium questions.

<center><h2> Conclusion </h2></center>

- The dataset is now prepared and cleaned for further analysis.
- Missing data is imputed and formatted correctly.
- Additional features are engineered for better insights.

In [50]:
# Saving the preprocessed DataFrame 'df'
# to a CSV file named 'preprocessed_data.csv' without including the index.
df.to_csv('preprocessed_data.csv', index=None)

In [51]:
pd.read_csv('preprocessed_data.csv').info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    3000 non-null   int64  
 1   page_number           3000 non-null   int64  
 2   is_premium            3000 non-null   bool   
 3   title                 3000 non-null   object 
 4   problem_description   2400 non-null   object 
 5   topic_tags            2400 non-null   object 
 6   difficulty            3000 non-null   object 
 7   similar_questions     2400 non-null   object 
 8   no_similar_questions  2400 non-null   float64
 9   acceptance            3000 non-null   float64
 10  accepted              2400 non-null   float64
 11  submission            2400 non-null   float64
 12  discussion_count      2399 non-null   float64
 13  likes                 2400 non-null   float64
 14  problem_URL           3000 non-null   object 
 15  solution_URL         